Source/ui_unix/GamePad/GamePadDeviceListener.cpp

```diff
@@ -0,0 +1,230 @@
+#include "GamePadDeviceListener.h"
+#include "GamePadUtils.h"
+#include <fcntl.h>
+#include <sys/signalfd.h>
+#include <poll.h>
+#include <sys/inotify.h>
+#include <csignal>
+
+#define EVENT_SIZE		( sizeof (struct inotify_event) )
+#define EVENT_BUF_LEN	( 1024 * ( EVENT_SIZE + NAME_MAX + 1) )
+#define WATCH_FLAGS		( IN_CREATE | IN_DELETE )
+
+CGamePadDeviceListener::CGamePadDeviceListener(OnInputEvent OnInputEventCallBack, bool filter)
+	: OnInputEventCallBack(OnInputEventCallBack),
+	  m_running(true),
+	  m_filter(filter)
+{
+
+	m_thread = std::thread(std::bind(&CGamePadDeviceListener::InputDeviceListenerThread, this));
+	UpdateDeviceList();
+
+}
+
+CGamePadDeviceListener::CGamePadDeviceListener(bool filter)
+	: m_filter(filter)
+{
+	m_thread = std::thread(std::bind(&CGamePadDeviceListener::InputDeviceListenerThread, this));
+	UpdateDeviceList();
+
+}
+
+CGamePadDeviceListener::~CGamePadDeviceListener()
+{
+	m_running = false;
+	m_GPIEList.clear();
+	if(m_thread.joinable()) m_thread.join();
```

Comment:
I initially though a thread could become unjoinable if it throw an exception, it seems i was mistaken, making the check redundant.